src/Parser/Mage/Frost/Modules/Features/BrainFreeze.js

```diff
@@ -11,81 +11,102 @@ const debug = false;
 // Brain Freeze appears to always fall after Flurry cast, but not always on same timestamp. Giving a margin here.
 const PROC_WINDOW_MS = 100;
 
+// When Glacial Spike is talented, it's OK to overwrite Brain Freeze procs if you have at least this many icicles during the frost bolt cast
+const ICICLE_BREAKPOINT = 3;
+
 class BrainFreeze extends Analyzer {
-	lastFlurryTimestamp;
+  lastFlurryTimestamp;
 
-	overwrittenProcs = 0;
-	expiredProcs = 0;
-	totalProcs = 0;
-	flurryWithoutProc = 0;
+  overwrittenProcs = 0;
+  okOverwrittenProcs = 0;
+  expiredProcs = 0;
+  totalProcs = 0;
+  flurryWithoutProc = 0;
 
-	on_byPlayer_applybuff(event) {
-		const spellId = event.ability.guid;
-    if (spellId !== SPELLS.BRAIN_FREEZE.id) {
-			return;
-		}
-		this.totalProcs += 1;
-	}
+  constructor(...args) {
+    super(...args);
+    this.glacialSpikeTalented = this.selectedCombatant.hasTalent(SPELLS.GLACIAL_SPIKE_TALENT.id);
+  }
+
+  on_byPlayer_applybuff(event) {
+    const spellId = event.ability.guid;
+    if (spellId === SPELLS.BRAIN_FREEZE.id) {
+      this.totalProcs += 1;
+    }
+  }
 
-	on_byPlayer_refreshbuff(event) {
-		const spellId = event.ability.guid;
+  on_byPlayer_refreshbuff(event) {
+    const spellId = event.ability.guid;
     if (spellId !== SPELLS.BRAIN_FREEZE.id) {
-			return;
-		}
-		this.overwrittenProcs += 1;
-		this.totalProcs += 1;
-    debug && console.log("Brain Freeze proc overwritten @ " + formatMilliseconds(event.timestamp - this.owner.fight.start_time));
+      return;
+    }
+    this.totalProcs += 1;
+
+    if (!this.glacialSpikeTalented) {
+      this.overwrittenProcs += 1;
+      debug && console.log("Brain Freeze proc overwritten w/o GS talented @ " + formatMilliseconds(event.timestamp - this.owner.fight.start_time));
+      return;
+    }
+
+    // Get the number of icicles the player had during the cast of the frostbolt
+    const stacksDuringCast = this.selectedCombatant.getBuff(SPELLS.ICICLES_BUFF.id, this.owner.currentTimestamp - 250).stacks || 0;
```

Comment:
Alright, so I took a look at the events and it looks like Icicles gains a stack right before frostbolt successfully completes its cast, therefore the stack count on getBuff when Frostbolt is cast will include the stack it just gained.

![image](https://user-images.githubusercontent.com/1304554/45602091-76fcf380-b9dd-11e8-8367-a3f32b04dfaa.png)

Also for confirmation I asked the theorycrafters in the Mage Discord and they mentioned that it is actually at 4 Icicles that you should hold Brain Freeze. and that if you have 3 Icicles and are in the middle of casting Frostbolt, that you should hold Brain Freeze cause you will have 4 Icicles after Frostbolt completes.

So i think we can drop the buffer and we should change the breakpoint to 4.

The confusion on the 3 or 4 icicles is probably my fault. I was told to change it to 4 and I did, and then i started getting complaints that wowanalyzer was saying 4 but the Altered Time guides were saying 3 so I changed it back to 3 cause thats what Altered Time said. Turns out the Altered Time guide was wrong and we should be using 4 not 3.

Reply:
Actually, scratch that. Looks like Brain Freeze is gained before Icicles. I should really reorder that, but leave the buffer in for now.

![image](https://user-images.githubusercontent.com/1304554/45602147-816bbd00-b9de-11e8-87b7-8b33666ce3a7.png)